bokehjs/src/coffee/core/util/eq.ts

```diff
@@ -0,0 +1,97 @@
+      return '' + a === '' + b
+    case '[object Number]':
+      // `NaN`s are equivalent, but non-reflexive.
+      // Object(NaN) is equivalent to NaN
+      if (+a !== +a) return +b !== +b
+      // An `egal` comparison is performed for other numeric values.
+      return +a === 0 ? 1 / +a === 1 / b : +a === +b
+    case '[object Date]':
+    case '[object Boolean]':
+      // Coerce dates and booleans to numeric primitive values. Dates are compared by their
+      // millisecond representations. Note that invalid dates with millisecond representations
+      // of `NaN` are not equivalent.
+      return +a === +b
+  }
+
+  const areArrays = className === '[object Array]'
+  if (!areArrays) {
+    if (typeof a != 'object' || typeof b != 'object') return false
+
+    // Objects with different constructors are not equivalent, but `Object`s or `Array`s
+    // from different frames are.
+    const aCtor = a.constructor, bCtor = b.constructor
+    if (aCtor !== bCtor && !(isFunction(aCtor) && aCtor instanceof aCtor &&
+                             isFunction(bCtor) && bCtor instanceof bCtor)
+                        && ('constructor' in a && 'constructor' in b)) {
+      return false
+    }
+  }
+  // Assume equality for cyclic structures. The algorithm for detecting cyclic
+  // structures is adapted from ES 5.1 section 15.12.3, abstract operation `JO`.
+
+  // Initializing stack of traversed objects.
+  // It's done here since we only need them for objects and arrays comparison.
+  aStack = aStack || []
+  bStack = bStack || []
+  let length = aStack.length
+  while (length--) {
+    // Linear search. Performance is inversely proportional to the number of
+    // unique nested structures.
+    if (aStack[length] === a) return bStack[length] === b
+  }
+
+  // Add the first object to the stack of traversed objects.
+  aStack.push(a)
+  bStack.push(b)
+
+  // Recursively compare objects and arrays.
+  if (areArrays) {
+    // Compare array lengths to determine if a deep comparison is necessary.
+    length = a.length
+    if (length !== b.length) return false
+    // Deep compare the contents, ignoring non-numeric properties.
+    while (length--) {
+      if (!eq(a[length], b[length], aStack, bStack)) return false
+    }
+  } else {
+    // Deep compare objects.
+    const keys = Object.keys(a)
+    let key
+    length = keys.length
+    // Ensure that both objects contain the same number of properties before comparing deep equality.
+    if (Object.keys(b).length !== length) return false
+    while (length--) {
+      // Deep compare each member
+      key = keys[length]
+      if (!(b.hasOwnProperty(key) && eq(a[key], b[key], aStack, bStack))) return false
+    }
+  }
+  // Remove the first object from the stack of traversed objects.
+  aStack.pop()
+  bStack.pop()
+  return true
+}
+
+// Perform a deep comparison to check if two objects are equal.
+export function isEqual(a: any, b: any) {
+  return eq(a, b)
+}
```

Comment:
Is this vendored/ported from something? If so a reference in a comment would be advised